core/include/gnuradio-4.0/Graph.hpp

```diff
@@ -591,16 +613,28 @@ class Edge {
     }
 };
 
-struct Graph {
+class Graph : public gr::Block<Graph> {
     alignas(hardware_destructive_interference_size) std::shared_ptr<gr::Sequence> progress                         = std::make_shared<gr::Sequence>();
     alignas(hardware_destructive_interference_size) std::shared_ptr<gr::thread_pool::BasicThreadPool> ioThreadPool = std::make_shared<gr::thread_pool::BasicThreadPool>(
             "graph_thread_pool", gr::thread_pool::TaskType::IO_BOUND, 2UZ, std::numeric_limits<uint32_t>::max());
 
 private:
     std::vector<std::function<ConnectionResult(Graph &)>> _connectionDefinitions;
-    std::vector<std::unique_ptr<BlockModel>>              _blocks;
     std::vector<Edge>                                     _edges;
 
+    std::vector<std::unique_ptr<BlockModel>> _blocks;
+
+    struct BlockData {
+        BlockModel                      *blockAccess = nullptr;
+        MsgPortOutNamed<"__ForChildren"> toChildMessagePort;
+        MsgPortInNamed<"__FromChildren"> fromChildMessagePort;
```

Comment:
Why not drop the nesting and then re-use the existing msgIn/out definitions from `Block<T>`?

Reply:
The separation is needed to differentiate between messages going in to those coming out of the Graph. If a block in a graph sends a message, we want it emitted to the graph's owner, not back to the Block.

Reply:
yes... forgot. These are the 'hidden' MsgPort that only the Graph uses to exchange data from/to it's children blocks. :+1: